vermouth/map_parser.py

```diff
@@ -0,0 +1,929 @@
+            `False` otherwise.
+        edge_match: collections.abc.Callable
+            A function that should take six arguments: two graphs, and four
+            node keys. The first two node keys will be in the first graph and
+            share an edge; and the last two node keys will be in the second
+            graph and share an edge. Should return `True` if a pair of edges
+            should be considered equal, and `False` otherwise.
+
+        Yields
+        ------
+        tuple[dict, vermouth.molecule.Block, dict]
+            A tuple containing 1) the correspondence between nodes in `graph`
+            and nodes in :attr:`blocks_to`, with the associated weights; 2)
+            :attr:`blocks_to`; and 3) :attr:`references` on which
+            :attr:`mapping` has been applied.
+        """
+        if node_match is None:
+            def node_match(node1, node2):
+                return True
+
+        if edge_match is None:
+            def edge_match(node11, node12, node21, node22):
+                return True
+        else:
+            edge_match = partial(edge_match, graph, self.blocks_from)
+
+        return self._graph_map(graph, node_match, edge_match)
+
+    def _graph_map(self, graph, node_match, edge_match):
+        """
+        Performs the partial mapping described in :meth:`map` using a
+        "classical" subgraph isomorphism algorithm.
+        """
+        # 1 Find subgraph isomorphism between blocks_from and graph
+        # 2 Translate found match ({graph idx: blocks from idx}) to indices in
+        #   blocks_to using self.mapping
+        # 3 Return found matches and blocks_to?
+        # PS. I don't really like this, because this object is becoming too
+        # intelligent by also having to do the isomorphism. On the other hand,
+        # it makes sense from the maths point of view.
+        matcher = MappingGraphMatcher(graph, self.blocks_from,
+                                      node_match=node_match,
+                                      edge_match=edge_match)
+        for match in matcher.subgraph_isomorphisms_iter():
+            rev_match = {v: k for k, v in match.items()}
+
+            new_match = defaultdict(dict)
+            for graph_idx, from_idx in match.items():
+                new_match[graph_idx].update(self.mapping[from_idx])
+            references = {out_idx: rev_match[ref_idx]
+                          for out_idx, ref_idx in self.references.items()}
+            yield dict(new_match), self.blocks_to, references
+
+    def _normalize_weights(self):
+        """
+        Normalize weights in :attr:`mapping` such that the sum of the weights
+        of nodes mapping to something is 1.
+        """
+        rev_mapping = self.reverse_mapping
+        for idx_from in self.mapping:
+            for idx_to, weight in self.mapping[idx_from].items():
+                self.mapping[idx_from][idx_to] = weight/sum(rev_mapping[idx_to].values())
+
+
+class MappingBuilder:
```

Comment:
Disagree. Also see my comment earlier. In addition, it prevents the logic of constructing a Mapping from cluttering it's API, and the Builder defines an interface that can be used by Directors to build Mappings. Currently there is 1 Director since there is 1 file format for mappings, but this is not necessarily true in the future.